File: BPM_week1.m
Nx = 128;
Ny = 128;
Nz = 128;
Nx_ext = 256;
Ny_ext = 256;
Nz_ext = 256;
nb = 1.333;
wl = 406;
drx = 100;
dry = 100;
drz = 100;
FPDist = -Nz*drz;
mod_dist=0;
dkx_ext = 2*pi/(Nx_ext*drx);
dky_ext = 2*pi/(Ny_ext*dry);
kx_ext = (dkx_ext * [0:Nx_ext/2-1, -Nx_ext/2:-1]');
ky_ext = (dky_ext * [0:Ny_ext/2-1, -Ny_ext/2:-1]');
[Kyy_ext,Kxx_ext] = meshgrid(ky_ext.^2, kx_ext.^2);
k = 2*pi/wl*nb;
k0 = 2*pi/wl; 
dphi_ext = real(k - sqrt(k^2 - Kxx_ext - Kyy_ext));% diffraction phase factor extended
x = nb* ones(Nx,Ny,Nz);
theta= 0;
u = zeros([Nx_ext,Ny_ext,Nz]);%Volume 3D du champ propagé
uin = ones(Nx_ext,Ny_ext);                
uouthat = uin;
                for ind_z = 1:Nz
             
                    uouthat = ifft2(fft2(uouthat).*exp(-1i*dphi_ext*drz));%Diffraction step
                   
                  
                    uouthat = uouthat.*padarray(exp(1i*k0*x(:,:,ind_z)*drz/cos(theta)),...
                        [(Nx_ext - Nx)/2,(Ny_ext - Ny)/2],1,'both');%Refraction step
                    u(:,:,ind_z) = uouthat;
                   
                end
                
                
                uouthat = ifft2(fft2(uouthat).*exp(-1i*dphi_ext*FPDist));%Free-space propagation until the focal plane
                
               
                uouthat = uouthat*exp(1i*k*mod_dist);%BPM original plan is missing this global phase, can be optional
                
                uouthat = uouthat(1 + end/2 - Nx/2:end/2 + Nx/2,1 + end/2 - Ny/2:end/2 + Ny/2);%Select the central image

File: BPM_week1.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Sep 18 12:00:38 2020

@author: paulmargain
"""
import numpy as np

###Initialisations
Nx, Ny, Nz = 128,128,128
Nx_ext, Ny_ext, Nz_ext = 256,256,256
nb = 1.333
wl = 406
pi=3.141592
i=complex(0,1)
drx, dry, drz = 100,100,100
FPDist = -Nz*drz
mod_dist=0
dkx_ext = 2*pi/(Nx_ext*drx)
dky_ext = 2*pi/(Ny_ext*dry)
kx_ext = dkx_ext * (np.array([k for k in range(Nx_ext//2)] + [ k for k in range (-Nx_ext//2,0)]))
ky_ext = dky_ext * (np.array([k for k in range(Ny_ext//2)] + [ k for k in range (-Ny_ext//2,0)]))
[Kyy_ext,Kxx_ext] = np.meshgrid(ky_ext**2, kx_ext**2)
k = 2*pi/wl*nb
k0 = 2*pi/wl        
dphi_ext = np.real(k - np.sqrt((k**2 - Kxx_ext - Kyy_ext),dtype="complex_")) #diffraction phase factor extended
x = nb* np.ones((Nx,Ny,Nz),dtype="complex_")
theta= 0;
u = np.zeros([Nx_ext,Ny_ext,Nz],dtype = "complex_") ##Volume 3D du champ propagé
uin = np.ones((Nx_ext,Ny_ext),dtype ="complex_")               
uouthat = uin
            
for ind_z in range (0,Nz):


    uouthat = np.fft.ifft2(np.fft.fft2(uouthat)*np.exp(-i*dphi_ext*drz)) #Diffraction step
    upad=uin
    upad[(Nx_ext-Nx)//2:(Nx_ext+Nx)//2,(Ny_ext-Ny)//2:(Ny_ext+Ny)//2] = np.exp(i*k0*x[:,:,ind_z]*drz/np.cos(theta))
    uouthat=uouthat*upad
    u[:,:,ind_z] = uouthat
    
                
               
uouthat = np.fft.ifft2(np.fft.fft2(uouthat)*np.exp(-i*dphi_ext*FPDist)) #Free-space propagation until the focal plane
                
        
uouthat = uouthat*np.exp(i*k*mod_dist) #BPM original plan is missing this global phase, can be optional


uouthat = uouthat[Nx//2:Nx_ext-Nx//2,Ny//2:Ny_ext-Ny//2 ] #Select the central image
       

